dm/worker/task_checker.go

```diff
@@ -0,0 +1,250 @@
+func NewRealTaskStatusChecker(cfg CheckerConfig, w *Worker) TaskStatusChecker {
+	tsc := &realTaskStatusChecker{
+		cfg: cfg,
+		l:   log.With(zap.String("component", "task checker")),
+		w:   w,
+	}
+	tsc.closed.Set(closedTrue)
+	return tsc
+}
+
+// Init implements TaskStatusChecker.Init
+func (tsc *realTaskStatusChecker) Init() error {
+	var err error
+	tsc.bf, err = backoff.NewBackoff(tsc.cfg.BackoffFactor, tsc.cfg.BackoffJitter, tsc.cfg.BackoffMin, tsc.cfg.BackoffMax)
+	return terror.WithClass(err, terror.ClassDMWorker)
+}
+
+// Start implements TaskStatusChecker.Start
+func (tsc *realTaskStatusChecker) Start() {
+	tsc.wg.Add(1)
+	go func() {
+		defer tsc.wg.Done()
+		tsc.run()
+	}()
+}
+
+// Close implements TaskStatusChecker.Close
+func (tsc *realTaskStatusChecker) Close() {
+	if !tsc.closed.CompareAndSwap(closedFalse, closedTrue) {
+		return
+	}
+
+	if tsc.cancel != nil {
+		tsc.cancel()
+	}
+	tsc.wg.Wait()
+}
+
+func (tsc *realTaskStatusChecker) run() {
+	tsc.ctx, tsc.cancel = context.WithCancel(context.Background())
+	tsc.normalTime = time.Now()
+	tsc.latestResume = time.Now()
+	tsc.closed.Set(closedFalse)
+	ticker := time.NewTicker(tsc.cfg.CheckInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-tsc.ctx.Done():
+			tsc.l.Info("worker task checker exits")
+			return
+		case <-ticker.C:
+			tsc.check()
+		}
+	}
+}
+
+// isRetryableError checks the error message and returns whether we need to
+// resume the task and retry
+func isRetryableError(err *pb.ProcessError) bool {
+	// not elegant code, because TiDB doesn't expose some error
+	unsupportedDDLMsgs := []string{
+		"can't drop column with index",
+		"unsupported add column",
+		"unsupported modify column",
+		"unsupported modify",
+		"unsupported drop integer primary key",
+	}
+	if err.Type == pb.ErrorType_ExecSQL {
+		for _, msg := range unsupportedDDLMsgs {
+			if strings.Contains(err.Msg, msg) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (tsc *realTaskStatusChecker) getResumeStrategy(taskStatus *pb.TaskStatus, duration time.Duration) ResumeStrategy {
+	// task that is not paused or paused manually, just ignore it
+	if taskStatus == nil || taskStatus.Stage != pb.Stage_Paused || taskStatus.Result == nil || taskStatus.Result.IsCanceled {
+		return ResumeIgnore
+	}
+
+	// TODO: use different strategies based on the error detail
+	for _, processErr := range taskStatus.Result.Errors {
+		if !isRetryableError(processErr) {
+			return ResumeNoSense
+		}
+	}
+
+	// auto resume interval does not exceed backoff duration, skip this paused task
+	if time.Since(tsc.latestResume) < duration {
+		return ResumeSkip
+	}
+
+	return ResumeDispatch
+}
+
+func (tsc *realTaskStatusChecker) check() {
+	tasks := tsc.w.getPausedSubTasks()
+	abnormal := 0
+	resumed := 0
+	duration := tsc.bf.Current()
+	for _, task := range tasks {
+		strategy := tsc.getResumeStrategy(task, duration)
+		if strategy == ResumeIgnore {
+			continue
+		}
+		abnormal++
```

Comment:
shall we move this line after L222? if had a lot of strategy is ResumeNoSense, the abnormal will always increase.